src/software/ai/hl/stp/tactic/intercept_ball/intercept_ball_fsm.h

```diff
@@ -0,0 +1,125 @@
+
+            if (robot_time_to_pos < ball_time_to_position)
+            {
+                break;
+            }
+            intercept_position +=
+                ball.velocity().normalize(INTERCEPT_POSITION_SEARCH_INTERVAL);
+        }
+        return intercept_position;
+    }
+
+    auto operator()()
+    {
+        using namespace boost::sml;
+
+        const auto chase_ball_s = state<chase_ball>;
+        const auto block_ball_s = state<MoveFSM>;
+
+        // update_e is the _event_ that the InterceptBallFSM responds to
+        const auto update_e = event<Update>;
+
+        /**
+         * Guard that checks if the ball is slow
+         *
+         * @param event InterceptBallFSM::Update
+         *
+         * @return if the ball is slow
+         */
+        const auto slow_ball = [](auto event) {
+            static const double BALL_MOVING_SLOW_SPEED_THRESHOLD = 0.3;
+            return event.common.world.ball().velocity().length() <
+                   BALL_MOVING_SLOW_SPEED_THRESHOLD;
+        };
+
+        /**
+         * Guard that checks if the ball has been intercepted
+         *
+         * @param event InterceptBallFSM::Update
+         *
+         * @return if the ball has been intercepted
+         */
+        const auto intercepted = [](auto event) {
+            return event.common.robot.isNearDribbler(
+                event.common.world.ball().position());
+        };
+
+        /**
+         * Action to update the MoveIntent to chase the ball
+         *
+         * @param event InterceptBallFSM::Update
+         */
+        const auto chase_ball = [this](auto event) {
+            auto ball_position = event.common.world.ball().position();
+            auto face_ball_orientation =
+                (ball_position - event.common.robot.position()).orientation();
+            event.common.set_intent(std::make_unique<MoveIntent>(
+                event.common.robot.id(), ball_position, face_ball_orientation, 0,
+                DribblerMode::MAX_FORCE, BallCollisionType::ALLOW));
+        };
+
+        /**
+         * Action to update MoveFSM
+         *
+         * @param event FastBallInterceptFSM::Update event
+         * @param processEvent processes the Move::Update
+         */
+        const auto block_ball = [this](auto event,
+                                       back::process<MoveFSM::Update> processEvent) {
+            Point intercept_position =
+                fastInterceptionPoint(event.common.robot, event.common.world.ball(),
+                                      event.common.world.field());
+            MoveFSM::ControlParams control_params{
+                .destination = intercept_position,
+                .final_orientation =
+                    (-event.common.world.ball().velocity()).orientation(),
+                .final_speed = 0};
+
+            processEvent(MoveFSM::Update(control_params, event.common));
+        };
+
+        return make_transition_table(
```

Comment:
This transition table is super readable and easy to follow, really awesome work 🔥 FSMs will win us games